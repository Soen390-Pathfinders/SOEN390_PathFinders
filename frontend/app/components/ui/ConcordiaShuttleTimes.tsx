import { View, Text, StyleSheet } from "react-native";
import { useState, useEffect } from "react";
import { loyolaSchedule, sgwSchedule } from "../../data/shuttleSchedules";
import ShuttleTimesDisplay from "../ui/ShuttleTimesDisplay.tsx";

//takes the time as String and returns a number of minutes
export const timeToMinutes = (time: string): number => {
  const [hours, minutes] = time.split(":").map(Number);
  // Check if the time is invalid (hours must be between 0-23, minutes between 0-59)
  if (
    isNaN(hours) ||
    isNaN(minutes) ||
    hours < 0 ||
    hours >= 24 ||
    minutes < 0 ||
    minutes >= 60
  ) {
    return NaN;
  }

  return hours * 60 + minutes;
};

// Converting the current time to minutes for accurate shuttle departure calculations. returns the current time in minutes
export const getCurrentTimeInMinutes = (): number => {
  return timeToMinutes(
    new Date().toLocaleTimeString("en-US", { hour12: false })
  );
};

//Format time for display
export const formatTime = (minutes: number): string => {
  const hours =
    minutes >= 1440
      ? Math.floor((minutes - 1440) / 60)
      : Math.floor(minutes / 60);
  const mins =
    minutes >= 1440
      ? Math.floor((minutes - 1440) % 60)
      : Math.floor(minutes % 60);
  return `${hours}:${mins.toString().padStart(2, "0")}`;
};

//getnextDepartures filters out the next 3 departures from the list of departure for both campus
export const getNextDepartures = (currentMinutes: number) => {
  const upcomingLoyola = loyolaSchedule
    .filter((time) => {
      const timeInMinutes = timeToMinutes(time);
      return timeInMinutes > currentMinutes - 1;
    })
    // Display only the next 3 shuttle times as per design choice to avoid cluttering the UI
    .slice(0, 3);

  const upcomingSGW = sgwSchedule
    .filter((time) => {
      const timeInMinutes = timeToMinutes(time);
      return timeInMinutes > currentMinutes - 1;
    })
    .slice(0, 3);
  return { loyola: upcomingLoyola, sgw: upcomingSGW };
};

export default function ConcordiaShuttleTimes() {
  const [currentTime, setCurrentTime] = useState(getCurrentTimeInMinutes()); //Get the current time from the devie in minutes
  const [departures, setDepartures] = useState(getNextDepartures(currentTime)); //departures array containing the next 3 based on current time.

  //updates current time and departures every minute
  useEffect(() => {
    const interval = setInterval(() => {
      const newCurrentTime = getCurrentTimeInMinutes();
      setCurrentTime(newCurrentTime);
      setDepartures(getNextDepartures(newCurrentTime));
    }, 60000);

    return () => clearInterval(interval);
  }, []);

  return (
    <View style={styles.timecontainer}>
      <Text style={styles.title}>Bus Shuttle Departure Times</Text>
      <Text style={styles.currentTime}>
        Current Time: {formatTime(currentTime)}
      </Text>

      <View style={styles.headerRow}>
        <Text style={styles.headerText}>LOYOLA CAMPUS</Text>
        <Text style={styles.headerText}>SGW CAMPUS</Text>
      </View>

      {[
        ...Array(Math.max(departures.loyola.length, departures.sgw.length)),
      ].map((_, index) => {
        const loyTime = departures.loyola[index];
        const sgwTime = departures.sgw[index];
        const loyMinutesAway = loyTime
          ? timeToMinutes(loyTime) - currentTime
          : null;
        const sgwMinutesAway = sgwTime
          ? timeToMinutes(sgwTime) - currentTime
          : null;

        return (
          <View key={index} style={styles.row}>
            <ShuttleTimesDisplay
              campusTime={loyTime}
              minutesAway={loyMinutesAway}
            />
            <ShuttleTimesDisplay
              campusTime={sgwTime}
              minutesAway={sgwMinutesAway}
            />
          </View>
        );
      })}
    </View>
  );
}

const styles = StyleSheet.create({
  timecontainer: {
    backgroundColor: "rgba(145, 35, 55, 0.99)",
    width: "100%",
    maxHeight: "37%",
    padding: 15,
  },
  title: {
    color: "white",
    fontSize: 18,
    fontWeight: "bold",
    textAlign: "center",
    marginBottom: 8,
    textTransform: "uppercase",
  },
  currentTime: {
    color: "white",
    fontSize: 16,
    textAlign: "center",
    marginBottom: 15,
  },
  headerRow: {
    flexDirection: "row",
    justifyContent: "space-around",
    marginBottom: 15,
    borderBottomWidth: 1,
    borderBottomColor: "white",
    paddingBottom: 10,
    paddingHorizontal: 20,
  },
  headerText: {
    color: "white",
    fontSize: 14,
    fontWeight: "bold",
    flex: 1,
    textAlign: "center",
  },
  row: {
    flexDirection: "row",
    justifyContent: "space-around",
    marginVertical: 8,
    paddingHorizontal: 20,
  },
  column: {
    flex: 1,
    alignItems: "center",
  },
  timeText: {
    color: "white",
    fontSize: 14,
    fontWeight: "500",
    textAlign: "center",
    minWidth: 150,
    flexDirection: "row",
    justifyContent: "center",
  },
  statusText: {
    fontSize: 14,
    marginLeft: 8,
    textAlign: "left",
    flex: 1,
  },
  departingStatus: {
    color: "#ffd700", // Gold color for departing
    fontWeight: "bold",
  },
  departedStatus: {
    color: "#48bb78", // Green color for departed
    fontWeight: "bold",
  },
  awayStatus: {
    color: "white", // Light blue for minutes away
  },
});
