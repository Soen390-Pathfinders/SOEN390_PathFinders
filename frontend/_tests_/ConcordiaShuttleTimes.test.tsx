import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react-native";
import ConcordiaShuttleTimes from "@/app/components/ui/ConcordiaShuttleTimes";
import {
  timeToMinutes,
  getCurrentTimeInMinutes,
  getNextDepartures,
  formatTime,
} from "@/app/components/ui/ConcordiaShuttleTimes";

//timeToMinutes function
describe("timeToMinutes", () => {
  test("converts a string of the form 10:00 to the number of minutes since midnights", () => {
    expect(timeToMinutes("10:00")).toBe(600);
  });

  test("converts a string of the form 0:00 to the number of minutes since midnights", () => {
    expect(timeToMinutes("00:00")).toBe(0);
  });
  test("converts 12:00 to 720 minutes", () => {
    expect(timeToMinutes("12:00")).toBe(720);
  });
  test("returns NaN for invalid time format", () => {
    expect(timeToMinutes("invalid")).toBeNaN();
  });
  test("returns NaN for out-of-range hours", () => {
    expect(timeToMinutes("25:30")).toBeNaN();
  });
});

//getCurrentTimeInMinutes function
describe("getCurrentTimeInMinutes", () => {
  //first of all mock the date object
  beforeAll(() => {
    //The fixed time is set to 17:00
    jest.spyOn(global, "Date").mockImplementation(() => {
      return {
        toLocaleTimeString: () => "17:00", // Mocked time
      };
    });
  });

  afterAll(() => {
    // Restore the original implementation of Date after tests are done
    jest.restoreAllMocks();
  });

  test("returns the mocked current time in minutes", () => {
    // will pass 17:00 to timeToMinutes function
    expect(getCurrentTimeInMinutes()).toBe(1020);
  });
});

//test getDepartures function
describe("getNextDepartures", () => {
  test("returns the next 3 departures after 17:01", () => {
    const currentMinutes = 1021; // 17:01
    expect(getNextDepartures(currentMinutes)).toEqual({
      loyola: ["17:15", "17:30", "17:45"],
      sgw: ["17:15", "17:30", "17:45"],
    });
  });
  test("returns the next 3 departures after 00:00", () => {
    const currentMinutes = 0; // 00:00
    expect(getNextDepartures(currentMinutes)).toEqual({
      loyola: ["9:15", "9:30", "9:45"],
      sgw: ["9:30", "9:45", "10:00"],
    });
  });
});
//formatTime
describe("formattime", () => {
  test("Formats the time from minutes to  a string of the form 0:00", () => {
    expect(formatTime(600)).toBe("10:00");
  });
  test("Formats the time from minutes to  a string of the form 0;00", () => {
    expect(formatTime(0)).toBe("0:00");
  });
  test("Formats the time from minutes to  a string of the form 0:00", () => {
    expect(formatTime(720)).toBe("12:00");
  });
  test("Formats the time from minutes to  a string of the form 0;00", () => {
    expect(formatTime(1441)).toBe("0:01");
  });
});

//test that the components renders correctly
describe("ConcordiaShuttleTimes rendering", () => {
  test("renders the shuttle departure times screen correctly", () => {
    // Renter the parent component
    render(<ConcordiaShuttleTimes />);

    //Make sure the title "Bus Shuttle Departure Times" is rendered
    expect(screen.getByText("Bus Shuttle Departure Times")).toBeTruthy();

    // Make sure the current time is rendering
    expect(screen.getByText(/Current Time:/)).toBeTruthy();

    // Are the row header rendering correctly
    expect(screen.getByText("LOYOLA CAMPUS")).toBeTruthy();
    expect(screen.getByText("SGW CAMPUS")).toBeTruthy();
  });
});

//This snapshot was removed because the date needs to be mocked or it will fail everytime -- we will rely less on snapshots form now on
//Add snapshot for Integration testing - making sure that nothing changes the rendering of the component in the future

describe("ConcordiaShuttleTimes", () => {
  beforeEach(() => {
    // Mock the date so the snapshot always compares the same text
    const mockDate = new Date(2025, 3, 14, 17, 0, 0); // 2025-04-14 17:00:00
    jest.spyOn(global, "Date").mockImplementation(() => mockDate);
  });

  test("renders the shuttle departure times screen correctly", () => {
    const { getByText } = render(<ConcordiaShuttleTimes />);

    // Check if some text or element is rendered to ensure the component renders
    expect(getByText("Bus Shuttle Departure Times")).toBeTruthy();
  });
  test("matches snapshot", () => {
    const { toJSON } = render(<ConcordiaShuttleTimes />);

    // // Create and compare snapshot
    expect(toJSON()).toMatchSnapshot();
  });
  afterEach(() => {
    // restore the date
    jest.restoreAllMocks();
  });
});
