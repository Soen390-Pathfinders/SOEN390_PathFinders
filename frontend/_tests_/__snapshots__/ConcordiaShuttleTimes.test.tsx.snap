// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConcordiaShuttleTimes matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(145, 35, 55, 0.99)",
      "maxHeight": "37%",
      "padding": 15,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 18,
        "fontWeight": "bold",
        "marginBottom": 8,
        "textAlign": "center",
        "textTransform": "uppercase",
      }
    }
  >
    Bus Shuttle Departure Times
  </Text>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 16,
        "marginBottom": 15,
        "textAlign": "center",
      }
    }
  >
    Current Time: 
    17:00
  </Text>
  <View
    style={
      {
        "borderBottomColor": "white",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginBottom": 15,
        "paddingBottom": 10,
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "flex": 1,
          "fontSize": 14,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      LOYOLA CAMPUS
    </Text>
    <Text
      style={
        {
          "color": "white",
          "flex": 1,
          "fontSize": 14,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      SGW CAMPUS
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginVertical": 8,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:00
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "#ffd700",
                "fontWeight": "bold",
              },
            ]
          }
        >
            ⏳ Departing...
        </Text>
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:00
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "#ffd700",
                "fontWeight": "bold",
              },
            ]
          }
        >
            ⏳ Departing...
        </Text>
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginVertical": 8,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:15
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "white",
              },
            ]
          }
        >
            🚌 in 15 min
        </Text>
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:15
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "white",
              },
            ]
          }
        >
            🚌 in 15 min
        </Text>
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginVertical": 8,
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:30
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "white",
              },
            ]
          }
        >
            🚌 in 30 min
        </Text>
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flexDirection": "row",
            "fontSize": 14,
            "fontWeight": "500",
            "justifyContent": "center",
            "minWidth": 150,
            "textAlign": "center",
          }
        }
      >
        17:30
        <Text
          style={
            [
              {
                "flex": 1,
                "fontSize": 14,
                "marginLeft": 8,
                "textAlign": "left",
              },
              {
                "color": "white",
              },
            ]
          }
        >
            🚌 in 30 min
        </Text>
      </Text>
    </View>
  </View>
</View>
`;
